VAULT-22483: Audit `filter` docs (#24903)
hashicorp/vault: website/content/docs/concepts/filtering/index.mdx

Author: this file is a port of the Boundary version (https://developer.hashicorp.com/boundary/docs/concepts/filtering)

---
layout: docs
page_title: Filtering
description: |-
  An introduction to the filtering syntax used in Vault.
---

# Filter expressions

Filter expressions can be used by various parts of Vault to provide useful
functionality. This page describes the overall syntax; see the sub-pages for
information on the specific values available for filtering for various features,
including examples.

Each filter expression has matching operators composed with selectors and
values.

## Create expressions

A single expression is a matching operator with a selector and value. They are
written in plain text format, and Boolean logic and parenthesization are
supported. In general whitespace is ignored, except within literal strings.

> ### Matching operators

Reviewer: Would it be possible to add some examples throughout, or maybe a big example at the end that puts all these things together?

Author: This page has been taken from Boundary in order for Vault to not have to reference their page in our docs.

https://developer.hashicorp.com/boundary/docs/concepts/filtering

---
layout: docs
page_title: Filtering
description: |-

>   An introduction to the filtering syntax used in Vault.

Author: It also already applied to Events https://developer.hashicorp.com/vault/docs/concepts/events

### Matching operators

All matching operators use a selector or value to choose what data should be
matched. Each endpoint that supports filtering accepts a potentially different
list of selectors and is detailed in the API documentation for those endpoints.

```text
// Equality & Inequality checks
<Selector> == "<Value>"
<Selector> != "<Value>"

// Emptiness checks
<Selector> is empty
<Selector> is not empty

// Contains checks or Substring Matching
"<Value>" in <Selector>
"<Value>" not in <Selector>
<Selector> contains "<Value>"
<Selector> not contains "<Value>"

// Regular Expression Matching
<Selector> matches "<Value>"
<Selector> not matches "<Value>"
```

### Selectors

Selectors are used by matching operators to create an expression. Inputs to
filter expressions are JSON (or JSON-derived); selectors therefore use [JSON
Pointer](https://tools.ietf.org/html/rfc6901) to select values from the input.
Each selector must be enclosed in quotes and contain a valid JSON Pointer path,
including leading slash (`/`).

```text
// Selects the value `zipzap` from the input `{ "foo": { "bar": "zipzap" } }`
"/foo/bar"
```

### Values

> Values are used by matching operators to create an expression. Values can be any
> valid selector, a number, or a string. It is best practice to quote values.
> 
> Numbers can be base 10 integers or floating point numbers.
> 
> When quoting strings, they may either be enclosed in double quotes or backticks.
> When enclosed in backticks they are treated as raw strings and escape sequences
> such as `\n` will not be expanded.

Reviewer 2: ```suggestion

Selector values can be any valid selector, integer, floating point number, or
string. Numbers and strings should be quoted in double quotes or backticks.

Strings quoted in backticks are treated as literal values and escape sequences
like `\n` are not expanded.

Value             | Type    | Expanded value
----------------- | ------- | ------------------------------ 
"Vault\tBoundary" | string  | "Vault	Boundary"
`Vault\tBoundary` | string  | "Vault\tBoundary"
"10"              | integer | "10"
`10`              | integer | "10"
"0.75"            | float   | "0.75"

```

> ## Connect expressions

Reviewer 2: ```suggestion
## Complex expressions
```

Suggest using "complex" expressions rather than "connecting" since there are compositions (like not) that may only include a single expression.

> There are several methods for connecting expressions, including
> 
> - Logical `or`
> - Logical `and`
> - Logical `not`
> - Grouping with parenthesis
> - Matching expressions

Reviewer 2: ```suggestion
Complex expressions combine basic expressions with logical operators, grouping, and matching expressions.
```

Suggest simplifying since you go into detail with the examples anyway

Reviewer 2: Follow up question: what, exactly, is a "matching expression". The syntax example doesn't really make that clear. From context I'm guessing a matching expression is a regex? If so, are there limitations to the types of regex we support? Or a specific syntax we can reference?

Author: Matching expressions are things like:

* `==`
* `!=`
* `is`
* `in`
* `contains`
* `matches` <-- this would be a regex

It comes from the HashiCorp bexpr library, which isn't well documented (there's a README but it doesn't break down all these expressions).

```text
// Logical Or - evaluates to true if either sub-expression does
<Expression 1> or <Expression 2>

// Logical And - evaluates to true if both sub-expressions do
<Expression 1 > and <Expression 2>

// Logical Not - evaluates to true if the sub-expression does not
not <Expression 1>

// Grouping - Overrides normal precedence rules
( <Expression 1> )

// Inspects data to check for a match
<Matching Expression 1>
```

> Standard operator precedence can be expected for the various forms. For example,
> the following two expressions would be equivalent.
> 
> ```text
> <Expression 1> and not <Expression 2> or <Expression 3>
> 
> ( <Expression 1> and (not <Expression 2> )) or <Expression 3>

Reviewer 2: ```suggestion
Vault uses standard operator precedence when  resolving complex
expressions. For example, the expression
`<Expression 1> and not <Expression 2> or <Expression 3>` resolves
the same as
`( <Expression 1> and (not <Expression 2> )) or <Expression 3>`.
```

Style correction: write in active voice

```

## Performance

> Filters are executed on Vault nodes and therefore will consume some amount
> of CPU time on the node.

Reviewer 2: ```suggestion
Filters consume a portion  of CPU time on the Vault node where they run.
```

Style correction: write in active voice

> <Note title="Regular expressions">

Reviewer 2: Performance question: comparing simple to complex is obvious, but what about a complex expression that joins multiple simple filters versus a single regex that captures the same thing?

Author: We know that regexes are more complicated and so will take more time, but we don't have public data on the trade offs in mixing and matching etc. Below this we try to allude to customers being encouraged to test things out beforehand to get an idea for their workloads.

  Using multiple/complex expressions including regular expressions
  (regex) will have a larger impact on performance than fewer/simpler filters.
</Note>

> We recommend that all filters are tested in pre-production environments in order
> to ensure correctness.

Reviewer 2: ```suggestion
Always test your filters in pre-production environments to ensure correctness.
```

Style correction: make best practices declarative.

Ideally you should [codify your management of Vault](/vault/tutorials/operations/codify-mgmt-vault-terraform)
using tools such as [Terraform](https://www.terraform.io/), to prevent accidentally enabling an audit device

> in a production environment with untested/incorrect settings.

Reviewer 2: What does this have to do with filtering? o_O

Author: Using Terraform for the management of Vault config is a best practice (our Tao speaks to codification).

We felt it was important to have this here as a reminder/guide to users to help reduce potential issues.

Especially since the audit configuration is immutable (meaning turning it off and on can render prior HMAC values incomparable), and can possibly block Vault if configured incorrectly.

With this in mind, would it be OK to leave this in please?

Finally, always ensure you profile production-like workloads within your pre-production

> environments in order to accurately assess the performance of Vault.

Reviewer 2: Same question here. The last two points seem like general administrative/performance advice, not anything that's specific to filtering. I'm not saying it isn't potentially useful, just that this doesn't seem to be the right place for it.

Author: This is because filtering is an advanced topic, and we want Operators to ensure they've profiled workloads on non-prod setups before going to prod. If they've done something like incorrectly configured lots of filters, this could lead to Vault trying to audit a lot and causing performance degradation. So we want to make sure we've given a friendly reminder that when you're in the world of filtering, you should be doing things by the book.
